File: src/core/piece.py
"""Base chess piece with battle stats."""
from __future__ import annotations
from typing import List, TYPE_CHECKING
from src.core.constants import WHITE, BLACK
from src.core.position import Position
from src.pieces.stats import PieceStats, default_stats

if TYPE_CHECKING:
    from src.core.board import Board


class Piece:
    symbol_map = {}  # subclasses register here

    def __init__(self, team: str, piece_type: str):
        self.team       = team          # WHITE or BLACK
        self.piece_type = piece_type
        self.has_moved  = False
        self.stats: PieceStats = default_stats(piece_type)

    # ------------------------------------------------------------------
    # Movement
    # ------------------------------------------------------------------

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        """Return all pseudo-legal destination squares."""
        raise NotImplementedError

    def _filter_valid(self, moves: List[Position], board: "Board") -> List[Position]:
        return [m for m in moves if m.is_valid()]

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def is_enemy(self, other: "Piece") -> bool:
        return other is not None and other.team != self.team

    def symbol(self) -> str:
        sym = self.symbol_map.get(self.piece_type, "?")
        return sym.upper() if self.team == WHITE else sym.lower()

    def __repr__(self) -> str:
        return f"{self.team[0]}{self.piece_type[:2]}"


# ---------------------------------------------------------------------------
# Concrete pieces
# ---------------------------------------------------------------------------

class Pawn(Piece):
    def __init__(self, team: str):
        from src.core.constants import PAWN
        super().__init__(team, PAWN)
        Piece.symbol_map[PAWN] = "P"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        moves = []
        direction = -1 if self.team == WHITE else 1  # White moves up (decreasing row)
        start_row  = 6 if self.team == WHITE else 1

        # One step forward
        fwd = pos + Position(direction, 0)
        if fwd.is_valid() and board.get(fwd) is None:
            moves.append(fwd)
            # Two steps from starting rank
            if not self.has_moved and pos.row == start_row:
                fwd2 = pos + Position(2 * direction, 0)
                if board.get(fwd2) is None:
                    moves.append(fwd2)

        # Captures (diagonal)
        for dc in (-1, 1):
            cap = pos + Position(direction, dc)
            if cap.is_valid() and self.is_enemy(board.get(cap)):
                moves.append(cap)

        return moves


class Rook(Piece):
    def __init__(self, team: str):
        from src.core.constants import ROOK
        super().__init__(team, ROOK)
        Piece.symbol_map[ROOK] = "R"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        return _slide(self, pos, board, [(1,0),(-1,0),(0,1),(0,-1)])


class Knight(Piece):
    def __init__(self, team: str):
        from src.core.constants import KNIGHT
        super().__init__(team, KNIGHT)
        Piece.symbol_map[KNIGHT] = "N"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        offsets = [(-2,-1),(-2,1),(-1,-2),(-1,2),(1,-2),(1,2),(2,-1),(2,1)]
        moves = []
        for dr, dc in offsets:
            dest = pos + Position(dr, dc)
            if dest.is_valid():
                target = board.get(dest)
                if target is None or self.is_enemy(target):
                    moves.append(dest)
        return moves


class Bishop(Piece):
    def __init__(self, team: str):
        from src.core.constants import BISHOP
        super().__init__(team, BISHOP)
        Piece.symbol_map[BISHOP] = "B"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        return _slide(self, pos, board, [(1,1),(1,-1),(-1,1),(-1,-1)])


class Queen(Piece):
    def __init__(self, team: str):
        from src.core.constants import QUEEN
        super().__init__(team, QUEEN)
        Piece.symbol_map[QUEEN] = "Q"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        return _slide(self, pos, board,
                      [(1,0),(-1,0),(0,1),(0,-1),(1,1),(1,-1),(-1,1),(-1,-1)])


class King(Piece):
    def __init__(self, team: str):
        from src.core.constants import KING
        super().__init__(team, KING)
        Piece.symbol_map[KING] = "K"

    def legal_moves(self, pos: Position, board: "Board") -> List[Position]:
        moves = []
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr == 0 and dc == 0:
                    continue
                dest = pos + Position(dr, dc)
                if dest.is_valid():
                    target = board.get(dest)
                    if target is None or self.is_enemy(target):
                        moves.append(dest)
        return moves


# ---------------------------------------------------------------------------
# Sliding helper
# ---------------------------------------------------------------------------

def _slide(piece: Piece, pos: Position, board: "Board",
           directions: list) -> List[Position]:
    moves = []
    for dr, dc in directions:
        cur = pos + Position(dr, dc)
        while cur.is_valid():
            target = board.get(cur)
            if target is None:
                moves.append(cur)
            elif piece.is_enemy(target):
                moves.append(cur)
                break
            else:
                break
            cur = cur + Position(dr, dc)
    return moves

File: web/js/core/piece.js
import { Team, PieceType } from "./constants.js";

export class Piece {
  constructor(team, type) {
    this.team     = team;
    this.type     = type;
    this.hasMoved = false;
  }

  isEnemy(other) { return other !== null && other.team !== this.team; }

  legalMoves(pos, board) { return []; }

  clone() {
    const p = new this.constructor(this.team);
    p.hasMoved = this.hasMoved;
    return p;
  }
}

// ── helpers ────────────────────────────────────────────────────────────────

function slide(piece, pos, board, dirs) {
  const moves = [];
  for (const [dr, dc] of dirs) {
    let [r, c] = [pos[0] + dr, pos[1] + dc];
    while (r >= 0 && r < 8 && c >= 0 && c < 8) {
      const target = board.get(r, c);
      if (target === null)       { moves.push([r, c]); }
      else if (piece.isEnemy(target)) { moves.push([r, c]); break; }
      else                       { break; }
      r += dr; c += dc;
    }
  }
  return moves;
}

// ── concrete pieces ────────────────────────────────────────────────────────

export class Pawn extends Piece {
  constructor(team) { super(team, PieceType.PAWN); }
  legalMoves([row, col], board) {
    const moves = [];
    const dir      = this.team === Team.WHITE ? -1 : 1;
    const startRow = this.team === Team.WHITE ?  6 : 1;

    const fwd = [row + dir, col];
    if (fwd[0] >= 0 && fwd[0] < 8 && board.get(...fwd) === null) {
      moves.push(fwd);
      const fwd2 = [row + 2 * dir, col];
      if (!this.hasMoved && row === startRow && board.get(...fwd2) === null)
        moves.push(fwd2);
    }
    for (const dc of [-1, 1]) {
      const cap = [row + dir, col + dc];
      if (cap[0] >= 0 && cap[0] < 8 && cap[1] >= 0 && cap[1] < 8 &&
          this.isEnemy(board.get(...cap)))
        moves.push(cap);
    }
    return moves;
  }
}

export class Rook extends Piece {
  constructor(team) { super(team, PieceType.ROOK); }
  legalMoves(pos, board) { return slide(this, pos, board, [[1,0],[-1,0],[0,1],[0,-1]]); }
}

export class Knight extends Piece {
  constructor(team) { super(team, PieceType.KNIGHT); }
  legalMoves([row, col], board) {
    return [[-2,-1],[-2,1],[-1,-2],[-1,2],[1,-2],[1,2],[2,-1],[2,1]]
      .map(([dr,dc]) => [row+dr, col+dc])
      .filter(([r,c]) => r>=0&&r<8&&c>=0&&c<8 &&
              (board.get(r,c)===null || this.isEnemy(board.get(r,c))));
  }
}

export class Bishop extends Piece {
  constructor(team) { super(team, PieceType.BISHOP); }
  legalMoves(pos, board) { return slide(this, pos, board, [[1,1],[1,-1],[-1,1],[-1,-1]]); }
}

export class Queen extends Piece {
  constructor(team) { super(team, PieceType.QUEEN); }
  legalMoves(pos, board) {
    return slide(this, pos, board, [[1,0],[-1,0],[0,1],[0,-1],[1,1],[1,-1],[-1,1],[-1,-1]]);
  }
}

export class King extends Piece {
  constructor(team) { super(team, PieceType.KING); }
  legalMoves([row, col], board) {
    const moves = [];
    for (const dr of [-1,0,1]) for (const dc of [-1,0,1]) {
      if (dr===0&&dc===0) continue;
      const [r,c] = [row+dr, col+dc];
      if (r>=0&&r<8&&c>=0&&c<8 && (board.get(r,c)===null || this.isEnemy(board.get(r,c))))
        moves.push([r,c]);
    }
    return moves;
  }
}
